src/node_large_page.cc

```diff
@@ -0,0 +1,255 @@
+    int
+      __attribute__((__section__(".eh_frame")))
+      __attribute__((__aligned__(2 * 1024 * 1024)))
+      __attribute__((__noinline__))
+      __attribute__((__optimize__("2")))
+      move_text_region_to_large_pages(struct TextRegion r) {
+        void *nmem = nullptr, *tmem = nullptr;
+        int ret = 0;
+
+        size_t size = (intptr_t)r.to - (intptr_t)r.from;
+        void *start = r.from;
+
+        // Allocate temporary region preparing for copy
+        nmem = mmap(nullptr, size,
+                    PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+        if (nmem == MAP_FAILED) {
+          printSystemError(errno);
+          return -1;
+        }
+
+        memcpy(nmem, r.from, size);
+
+// We already know the original page is r-xp
+// (PROT_READ, PROT_EXEC, MAP_PRIVATE)
+// We want PROT_WRITE because we are writing into it.
+// We want it at the fixed address and we use MAP_FIXED.
+        tmem = mmap(start, size,
+                    PROT_READ | PROT_WRITE | PROT_EXEC,
+                    MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1 , 0);
+        if (tmem == MAP_FAILED) {
+          printSystemError(errno);
+          munmap(nmem, size);
+          return -1;
+        }
+
+        if (tmem != start) {
+          fprintf(stderr, "Unable to allocate hugepages\n");
+          munmap(nmem, size);
+          munmap(tmem, size);
+          return -1;
+        }
+
+        ret = madvise(tmem, size, MADV_HUGEPAGE);
+        if (ret == -1) {
+          printSystemError(errno);
+          ret = munmap(tmem, size);
+          if (ret == -1) {
+            printSystemError(errno);
+          }
+          ret = munmap(nmem, size);
+          if (ret == -1) {
+            printSystemError(errno);
+          }
+
+          return -1;
+        }
+
+        memcpy(start, nmem, size);
+        ret = mprotect(start, size, PROT_READ | PROT_EXEC);
+        if (ret == -1) {
+          printSystemError(errno);
+          ret = munmap(tmem, size);
+          if (ret == -1) {
+            printSystemError(errno);
+          }
+          ret = munmap(nmem, size);
+          if (ret == -1) {
+            printSystemError(errno);
+          }
+          return -1;
+        }
+
+        // Release the old/temporary mapped region
+        ret = munmap(nmem, size);
+        if (ret == -1) {
+          printSystemError(errno);
+        }
+
+        return ret;
+    }
+
+    // This is the primary API called from main
+    int map_static_code_to_large_pages() {
+      struct TextRegion r = find_node_text_region();
+      if (r.found_text_region == false) {
+        fprintf(stderr, "Hugepages WARNING: failed to find text region \n");
+        return -1;
+      }
+
+      if (r.to <= reinterpret_cast<void *> (&move_text_region_to_large_pages))
+        return move_text_region_to_large_pages(r);
+
+      return -1;
+    }
+
+    bool isLargePagesEnabled() {
+      return isTransparentHugePagesEnabled();
+    }
+
+}  // namespace node
```

Comment:
I admit I kinda checked out after the first 100 lines.  This file has so many style errors, can you fix those first, please?

Reply:
Could you be specific? I ran the make -s lint and fixed all the ones pointed to by that.

Reply:
The linter doesn't catch everything but to name three things:

1. Inconsistent indentation.
2. Star leans left in pointers (`void* p`, not `void *p` or `void * p`.)
3. Casing: `IsLargePagesEnabled`, not `isLargePagesEnabled`.

Reply:
Thanks I have updated the PR with 2) and 3). I am using standard indentation of 2 spaces.

Reply:
@suresh-srinivas There’s https://github.com/nodejs/node/blob/master/CPP_STYLE_GUIDE.md that would help a lot with this. Also, comparing with other code in our codebase if you’re unsure. :)

(In particular, we don’t indent code inside namespaces)

Reply:
@addaleax thanks much. I will look at this and fix styles accordingly.

Reply:
@addaleax @bnoordhuis @gireeshpunathil fixed the style errors. Let me know if you see anything else.